TLS client management and other changes for clip upload in Push AV

raveendra-karu opened the pull request:
#### Summary
This PR takes care of following changes:
1. Using TLS client management delegate from camera device to get certificate, key details and utilizing for clip upload
2. Support for Chunk duration and Segment duration in milli seconds
3. Add GeneratePushTransportBeginEvent, GeneratePushTransportEndEvent for Clip record start and stop
4. Handle clip record based on Zone IDs and Sensitivity triggers
5. Add implementation for calculation of current bandwidth usage for clip record

#### Testing
```
# Build and run camera application
./scripts/examples/gn_build_example.sh examples/camera-app/linux out/debug/
sudo rm -rf /tmp/chip_*; ./out/debu/chip-camera-app

# Build python controller and active environment
scripts/build_python.sh -m platform -i out/python_env
source out/python_env/bin/activate

# Run Push AV integration TCs 

example:
python3 src/python_testing/TC_PAVSTI_2_2.py --commissioning-method on-network --qr-code MT:-24J0AFN00KA0648G00 --endpoint 1
```

raveendra-karu committed:
Initial changes for getting TLS client management delegate to PAVST delegate

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
- Added TLS certificates to Push AV transport manager.
- Updated zone management and triggers for PushAV.
- Updated zone ID validation logic.
- Implemented bandwidth limit API.
- Updated Push AV Stream Transport end event generation logic.

Co-author: Sambhavi <sambhavi.1@samsung.com>
Co-author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
- Replace file-based TLS certificates with in-memory buffer management
- Add SetOnRecorderStartedCallback functionality across transport classes
- Generate PushTransportBeginEvent when recorder starts streaming
- Update PushAVUploader to support both path and buffer-based certificates
- Refactor certificate structures for better type safety
- Add certificate table accessor methods for improved encapsulation

Co-author: Tushant Yadav <t.yadav@samsung.com>
Co-author: Sambhavi <sambhavi.1@samsung.com>

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Updated url handling and clip segment duration

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

marktrayer commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-logic.cpp:
```diff
@@ -542,35 +549,38 @@ PushAvStreamTransportServerLogic::HandleAllocatePushTransport(CommandHandler & h
+    // auto status = to_underlying(StatusCodeEnum::kInvalidTLSEndpoint);
+    // handler.AddClusterSpecificFailure(commandPath, status);
+    // return std::nullopt;
+    // }
```
Why this big block of commented out code?

yufengwangca commented on examples/camera-app/camera-common/src/camera-app.cpp:
```diff
@@ -50,6 +51,8 @@ CameraApp::CameraApp(chip::EndpointId aClustersEndpoint, CameraDeviceInterface *
                                                                     &Clusters::TlsClientManagementCommandDelegate::GetInstance());
 
+    Clusters::PushAvStreamTransport::SetTlsCertificateManagementDelegate(
+        chip::EndpointId(1), &Clusters::TlsCertificateManagementCommandDelegate::getInstance());
```
we should set to mEndpoint instead of hard-coded endpoint 1

yufengwangca commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-delegate.h:
```diff
@@ -242,6 +262,17 @@ class PushAvStreamTransportDelegate
+
+    virtual void
+    SetOnRecorderStartedCallback(std::function<void(uint16_t, PushAvStreamTransport::TransportTriggerTypeEnum)> cb) = 0;
+    virtual void SetTLSCerts(Tls::CertificateTable::BufferedClientCert clientCertEntry,
```
```suggestion
    virtual void SetTLSCerts(Tls::CertificateTable::BufferedClientCert & clientCertEntry,
```

yufengwangca commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-delegate.h:
```diff
@@ -242,6 +262,17 @@ class PushAvStreamTransportDelegate
+    virtual void
+    SetOnRecorderStartedCallback(std::function<void(uint16_t, PushAvStreamTransport::TransportTriggerTypeEnum)> cb) = 0;
+    virtual void SetTLSCerts(Tls::CertificateTable::BufferedClientCert clientCertEntry,
+                             Tls::CertificateTable::BufferedRootCert rootCertEntry)                                 = 0;
```
```suggestion
                             Tls::CertificateTable::BufferedRootCert & rootCertEntry)                                 = 0;
```

yufengwangca commented on examples/camera-app/linux/include/clusters/push-av-stream-transport/push-av-stream-manager.h:
```diff
@@ -69,6 +70,21 @@ class PushAvStreamTransportManager : public PushAvStreamTransportDelegate
         const Optional<Structs::TransportMotionTriggerTimeControlStruct::DecodableType> & timeControl) override;
 
+    void SetTLSCerts(Tls::CertificateTable::BufferedClientCert clientCertEntry,
+                     Tls::CertificateTable::BufferedRootCert rootCertEntry) override
```
```suggestion
                     Tls::CertificateTable::BufferedRootCert & rootCertEntry) override
```

yufengwangca commented on examples/camera-app/linux/include/clusters/push-av-stream-transport/push-av-stream-manager.h:
```diff
@@ -104,6 +126,24 @@ class PushAvStreamTransportManager : public PushAvStreamTransportDelegate
+    std::function<void(uint16_t, PushAvStreamTransport::TransportTriggerTypeEnum)> mOnRecorderStartedCb;
+
+    std::vector<uint8_t> bufferRootCert;
+    std::vector<uint8_t> bufferClientCert;
```
```suggestion
    std::vector<uint8_t> mBufferClientCert;
```

yufengwangca commented on examples/camera-app/linux/include/clusters/push-av-stream-transport/push-av-stream-manager.h:
```diff
@@ -69,6 +70,21 @@ class PushAvStreamTransportManager : public PushAvStreamTransportDelegate
+        bufferClientCert.assign(clientSpan.data(), clientSpan.data() + clientSpan.size());
+
+        auto * clientKeyPtr = clientCertEntry.mCertWithKey.key.Bytes();
+        auto clientKeyLen   = clientCertEntry.mCertWithKey.key.Length();
```
const auto & key = clientCertEntry.mCertWithKey.key;
bufferClientCertKey.assign(key.Bytes(), key.Bytes() + key.Length());

raveendra-karu committed:
Updated certificate handling in push AV stream transport with improved credential and crypto support.

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by clang-format

raveendra-karu committed:
Add recorder callbacks and refactor TLS cert handling

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Improved Null Checks and Certificate Buffer Allocation
1. Add null check for  in  to prevent null pointer dereference.
2. Replace stack-allocated certificate buffers with heap-allocated ones to reduce stack usage and handle large certificate buffers efficiently.

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Add override keywords to SetOnRecorderStoppedCallback and SetOnRecorderStartedCallback functions in push-av-stream-manager.h to properly implement interface methods.

raveendra-karu committed:
Fix TLS Key handling in push AV stream transport

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
resolved lint error

robszewczyk committed:
Merge branch 'master' into pushav/tls-support

robszewczyk commented:
/gemini review

pidarped commented on examples/camera-app/linux/include/clusters/push-av-stream-transport/push-av-stream-manager.h:
```diff
@@ -52,6 +55,8 @@ class PushAvStreamTransportManager : public PushAvStreamTransportDelegate
     void Init();
     void SetMediaController(MediaController * mediaController);
     void SetCameraDevice(CameraDeviceInterface * cameraDevice);
+    void SetOnRecorderStoppedCallback(std::function<void(uint16_t, PushAvStreamTransport::TransportTriggerTypeEnum)> cb) override;
```
It is not clear what the uint16_t arguments may represent. Please document above some of these details.

pidarped commented on examples/camera-app/linux/src/clusters/push-av-stream-transport/push-av-stream-manager.cpp:
```diff
@@ -222,18 +238,96 @@ Protocols::InteractionModel::Status PushAvStreamTransportManager::ManuallyTrigge
+    {
+        uint16_t vStreamId = videoStreamId.Value().Value();
+
+        auto & availableVideoStreams = mCameraDevice->GetCameraHALInterface().GetAvailableVideoStreams();
```
This needs to calculate against the allocated streams' bitrate configurations, not the available streams in the camera device.
So, towards that, this function should call into the CameraAVStreamMgmt delegate to calculat ethe stream bandwidth because it has access to the allocated streams.

pidarped commented on examples/camera-app/linux/src/clusters/push-av-stream-transport/push-av-stream-manager.cpp:
```diff
@@ -222,18 +238,96 @@ Protocols::InteractionModel::Status PushAvStreamTransportManager::ManuallyTrigge
+    {
+        uint16_t aStreamId = audioStreamId.Value().Value();
+
+        auto & availableAudioStreams = mCameraDevice->GetCameraHALInterface().GetAvailableAudioStreams();
```
Same for the audio stream

raveendra-karu committed:
Merge branch 'project-chip:master' into pushav/tls-support

raveendra-karu committed:
Add intermediate certificates support to TLS handling in PushAV

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Updated log

raveendra-karu committed:
Update PushAV event generation logic to avoid dependency inversion violation + other minor fixes

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Merge branch 'master' into pushav/tls-support

raveendra-karu committed:
Build fail fix

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Add GetAllocatedVideoStreams and GetAllocatedAudioStreams methods to CameraAVStreamManager class to provide direct access to allocated streams. Update PushAvStreamTransportManager to use the new delegate interface instead of accessing streams through HAL interface.

Author: Tushant Yadav <t.yadav@samsung.com>
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Add getters for allocated AV streams in CameraAVStreamManager

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-logic.h:
```diff
@@ -29,8 +31,41 @@ class PushAvStreamTransportServerLogic
             ChipLogError(Zcl, "Push AV Stream Transport : Trying to set delegate to null");
             return;
         }
+        mDelegate->SetPushAvStreamTransportServer(this);
```
Sure. Created an issue for this https://github.com/project-chip/connectedhomeip/issues/40980
I'll fix in the follow-up PR

raveendra-karu commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-logic.cpp:
```diff
@@ -596,6 +596,32 @@ PushAvStreamTransportServerLogic::HandleAllocatePushTransport(CommandHandler & h
         });
+        // Use heap allocation for large certificate buffers to reduce stack usage
+        auto rootCertBuffer   = std::make_unique<PersistentStore<CHIP_CONFIG_TLS_PERSISTED_ROOT_CERT_BYTES>>();
+        auto clientCertBuffer = std::make_unique<PersistentStore<CHIP_CONFIG_TLS_PERSISTED_CLIENT_CERT_BYTES>>();
```
Sure. Created an issue for this https://github.com/project-chip/connectedhomeip/issues/40980
I'll fix in the follow-up PR

raveendra-karu commented on examples/camera-app/linux/src/camera-device.cpp:
```diff
@@ -1374,6 +1374,7 @@ CameraError CameraDevice::RemoveZoneTrigger(const uint16_t zoneID)
 void CameraDevice::HandleSimulatedZoneTriggeredEvent(uint16_t zoneID)
 {
     mZoneManager.OnZoneTriggeredEvent(zoneID, ZoneEventTriggeredReasonEnum::kMotion);
+    mPushAVTransportManager.OnZoneTriggeredEvent(zoneID);
```
Sure. Created an issue for this https://github.com/project-chip/connectedhomeip/issues/40980
I'll fix in the follow-up PR

raveendra-karu commented on examples/camera-app/linux/include/pushav-clip-recorder.h:
```diff
@@ -138,6 +141,22 @@ class PushAVClipRecorder
+    void SetOnStopCallback(std::function<void()> cb) { mOnStopCallback = std::move(cb); }
+
+    // Set the cluster server reference for direct API calls
+    void SetPushAvStreamTransportServer(chip::app::Clusters::PushAvStreamTransportServerLogic * server)
```
Sure. Created an issue for this https://github.com/project-chip/connectedhomeip/issues/40980
I'll fix in the follow-up PR
